### DISCUSSION TITLE
Best way to install/deploy custom Go plugin?

### DISCUSSION THREAD

**Discussion by abork-dolby** [Kong/kong#7904] [Help] [answered]
What is the best way to **install/deploy** custom Go plugin in the Kong/KIC?

I was looking at [this documentation](https://docs.konghq.com/gateway-oss/2.5.x/external-plugins) but it only gives information how to build one. I chose to use Go plugin with embedded server. I also looked at [this](https://docs.konghq.com/gateway-oss/2.5.x/plugin-development/distribution/) but it seems like it covers Lua plugins. I tried to do similar thing but with Go and I ended up creating multi-stage docker build where I compile Go package and then copy it to custom Kong runtime image.

This worked fine when I used [this](https://hub.docker.com/layers/kong/go-plugin-tool/latest-alpine-latest/images/sha256-ee6c03d1944b3ba4a394b159e017a82da9932858efe390e60a4718a38b5bb94a?context=explore) as build image. Now, I wanted to bump up Go version (from `1.13.5`) and first I didn't find the source code of this image and then I tried to use official Go image as my build image but this didn't work.

My question is:
1) Is this even right way to install custom Go plugin?
2) Is there a source code for [this](https://hub.docker.com/layers/kong/go-plugin-tool/latest-alpine-latest/images/sha256-ee6c03d1944b3ba4a394b159e017a82da9932858efe390e60a4718a38b5bb94a?context=explore) image?

### Some info

Go plugin
```go
package main

import (
	"github.com/Kong/go-pdk"
	"github.com/Kong/go-pdk/server"
	"github.com/abork-dolby/kong-plugin-oidc-validate/jwt"
)

type Config struct {
	TrustedIss string
}

func New() interface{} {
	return &Config{TrustedIss: "default"}
}

func (conf Config) Access(kong *pdk.PDK) {
	verifier := jwt.JwtTokenVerifier{}

	authHeader, _ := kong.Request.GetHeader("Authorization")
	isValid := verifier.Verify(authHeader)
	if isValid == false {
		kong.Response.ExitStatus(401)
	}

	_ = kong.Log.Info("Plugin stopped")
}

func main() {
	Version := "1.1"
	Priority := 1000

	_ = server.StartServer(New, Version, Priority)
}
```

Docker file to build custom image "abork/kong"
```Docker
# This worked fine
# FROM kong/go-plugin-tool:2.0.4-alpine-latest AS build
# If I only change the image and leave everything else the same, kong proxy container fails in the runtime
FROM golang:buster AS build
RUN mkdir -p /custom-plugins/
COPY ./src /custom-plugins/
RUN cd /custom-plugins/ && \
   go build

FROM kong:2.5 as runtime
USER root
COPY --from=build /custom-plugins/kong-plugin-oidc-validate /usr/local/bin/oidc-validate
#COPY ./kong.yml /usr/local/kong/declarative/kong.yml
RUN chmod -R 777 /usr/local/bin/oidc-validate
USER kong
```

Helm values to install KIC on minikube
```yaml
env:
  LOG_LEVEL: info
  PLUGINS: oidc-validate
  PLUGINSERVER_NAMES: oidc-validate
  PLUGINSERVER_OIDC_VALIDATE_QUERY_CMD: /usr/local/bin/oidc-validate -dump

image:
  repository: abork/kong
  tag: "2.5-4"

proxy:
  type: NodePort

```

**Comment by javierguerragiraldez** [comment, marked as answer]
The `go-plugin-tool` image wasn't supposed to be used as a base image, it was an attempt to provide a packaged "version-appropriate" golang toolset to comply with the requirements to build runtime-linkable Go plugins.  Unfortunately, it didn't help to reduce confusion, we should remove it.

Much better is to avoid runtime-loadable code altogether, and that's why we went with the embedded server approach (which you're already using in your example).  That makes each plugin a separate process, and removes the need to match compiler version and environment with the main Kong package.

In short, this means you can freely use whatever version of Go that is appropriate for your target environment. `golang:alpine` or `golang:buster` might be OK.  the biggest thing to remember is that musl-based distros (like Alpine Linux) are incompatible with glibc-based ones (almost any other distro)

**Reply by borkke** [reply]
Hi Javier, thanks for the help! I'm glad that embedded server is the preferred way, it does look like a very neet solution. 

Just want to confirm one more thing. In order to get that plugin (compiled Go code) into Kong, is it safe to say that the preferred way is to take the official Kong image as a base image and create your own custom image for Kong that has my Go binary?

**Reply by javierguerragiraldez** [reply]
yes, when you write a plugin (in any language), you're adding code to your Kong deployment, so it's best to have your own image with it included
